source/exe/main_common.cc

```diff
@@ -89,6 +104,251 @@ void MainCommonBase::adminRequest(absl::string_view path_and_query, absl::string
+    request_ = opt_admin_->makeRequest(filter);
+    code_ = request_->start(*response_headers_);
+    {
+      absl::MutexLock lock(&mutex_);
+      if (headers_fn_ == nullptr || cancelled_) {
+        return;
+      }
+      Server::Utility::populateFallbackResponseHeaders(code_, *response_headers_);
+      headers_fn_(code_, *response_headers_);
+      headers_fn_ = nullptr;
+    }
+  }
+
+  void requestNextChunk() {
+    ASSERT_IS_MAIN_OR_TEST_THREAD();
+    {
+      absl::MutexLock lock(&mutex_);
+      if (cancelled_ || terminated_) {
+        return;
+      }
+    }
+    while (response_.length() == 0 && more_data_) {
+      more_data_ = request_->nextChunk(response_);
+    }
+    {
+      absl::MutexLock lock(&mutex_);
+      if (sent_end_stream_ || cancelled_) {
+        return;
+      }
+      sent_end_stream_ = !more_data_;
+      body_fn_(response_, more_data_);
+      ASSERT(response_.length() == 0);
+      body_fn_ = nullptr;
+    }
+  }
+
+  void sendAbortChunkLockHeld() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mutex_) {
+    if (!sent_end_stream_ && body_fn_ != nullptr) {
+      response_.drain(response_.length());
+      body_fn_(response_, false);
+      sent_end_stream_ = true;
+    }
+    body_fn_ = nullptr;
+  }
+
+  void sendErrorLockHeld() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mutex_) {
+    if (headers_fn_ != nullptr) {
+      code_ = Http::Code::InternalServerError;
+      Server::Utility::populateFallbackResponseHeaders(code_, *response_headers_);
+      headers_fn_(code_, *response_headers_);
+      headers_fn_ = nullptr;
+    }
+  }
+
+  Server::Instance& server_;
+  OptRef<Server::Admin> opt_admin_;
+  Buffer::OwnedImpl response_;
+  Http::Code code_;
+  Server::Admin::RequestPtr request_;
+  CleanupFn cleanup_;
+  Http::RequestHeaderMapPtr request_headers_{Http::RequestHeaderMapImpl::create()};
+  Http::ResponseHeaderMapPtr response_headers_{Http::ResponseHeaderMapImpl::create()};
+  bool more_data_ = true;
+
+  // True if cancel() was explicitly called by the user; headers and body
+  // callbacks are never called after cancel().
+  bool cancelled_ ABSL_GUARDED_BY(mutex_) = false;
+
+  // True if the Envoy server has stopped running its main loop. Headers and
+  // body requests can be initiated and called back are called after terminate,
+  // so callers do not have to special case this -- the request will simply fail
+  // with an empty response.
+  bool terminated_ ABSL_GUARDED_BY(mutex_) = false;
+
+  // Used to indicate whether the body function has been called with false
+  // as its second argument. That must always happen at most once, even
+  // if terminate races with the normal end-of-stream marker. more=false
+  // may never be sent if the request is cancelled, nor deleted prior to
+  // it being requested.
+  bool sent_end_stream_ ABSL_GUARDED_BY(mutex_) = false;
+
+  HeadersFn headers_fn_ ABSL_GUARDED_BY(mutex_);
+  BodyFn body_fn_ ABSL_GUARDED_BY(mutex_);
+  mutable absl::Mutex mutex_;
```

Comment:
Would it be possible (and make sense) to reuse the `mutex_` from the parent class (maybe shift it to protected from private)? Having two member mutexes named `mutex_` in the same class hierarchy seems like a really good setup for confusion and surprise deadlocks.